File: 백준/Gold/27440. 1로 만들기 3/1로 만들기 3.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.Map;
import java.util.Queue;

public class Main {
    private static int minCompute(long n) {
        if (n == 1) return 0;

        Queue<long[]> queue = new ArrayDeque<>();
        Map<Long, Integer> visited = new HashMap<>();

        queue.offer(new long[]{n, 0});
        visited.put(n, 0);

        while (!queue.isEmpty()) {
            long[] cur = queue.poll();
            long num = cur[0];
            int cnt = (int) cur[1];

            if (num == 1) {
                return cnt;
            }

            if (visited.containsKey(num) && visited.get(num) < cnt) {
                continue;
            }

            if (num % 3 == 0) {
                long next = num / 3;
                if (!visited.containsKey(next) || visited.get(next) > cnt + 1) {
                    visited.put(next, cnt + 1);
                    queue.offer(new long[]{next, cnt + 1});
                }
            }

            if (num % 2 == 0) {
                long next = num / 2;
                if (!visited.containsKey(next) || visited.get(next) > cnt + 1) {
                    visited.put(next, cnt + 1);
                    queue.offer(new long[]{next, cnt + 1});
                }
            }

            long next = num - 1;
            if (!visited.containsKey(next) || visited.get(next) > cnt + 1) {
                visited.put(next, cnt + 1);
                queue.offer(new long[]{next, cnt + 1});
            }
        }

        return -1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        long n = Long.parseLong(br.readLine());

        System.out.println(minCompute(n));
    }
}

File: 백준/Gold/27440. 1로 만들기 3/1로 만들기 3.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

private fun minCompute(n: Long): Int {
    if (n == 1L) return 0

    val queue: Queue<Pair<Long, Int>> = LinkedList()
    val visited = mutableMapOf<Long, Int>()

    queue.offer(n to 0)
    visited[n] = 0

    while (queue.isNotEmpty()) {
        val (num, cnt) = queue.poll()

        if (num == 1L) {
            return cnt
        }

        if (visited[num]!! < cnt) continue

        if (num % 3 == 0L) {
            val next = num / 3
            if (next !in visited || visited[next]!! > cnt + 1) {
                visited[next] = cnt + 1
                queue.offer(next to cnt + 1)
            }
        }

        if (num % 2 == 0L) {
            val next = num / 2
            if (next !in visited || visited[next]!! > cnt + 1) {
                visited[next] = cnt + 1
                queue.offer(next to cnt + 1)
            }
        }

        val next = num - 1
        if (next !in visited || visited[next]!! > cnt + 1) {
            visited[next] = cnt + 1
            queue.offer(next to cnt + 1)
        }
    }

    return -1
}

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toLong()

    println(minCompute(n))
}
